lib/evaluate/character.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_CHARACTER_H_
+#define FORTRAN_EVALUATE_CHARACTER_H_
+
+#include "type.h"
+#include <string>
+
+// character(1) is ISO/IEC 646:1991 (ASCII)
+//
+// character(2) is a variant of EUC-JP: The code points are the same as EUC-JP,
+// but the encoding has a two bytes fix-size (EUC-JP encoding has a variable
+// length). The one byte EUC-JP character representations are simply
+// zero-extended to two byte representations. The three bytes character
+// representation of EUC-JP (JIS X 0212) are not supported in this internal
+// encoding.
+//
+// character(4) is ISO/IEC 10646 UCS-4 (~ UTF-32)
+
+namespace Fortran::evaluate {
+
+template<int KIND> class CharacterUtils {
+  using Character = Scalar<Type<TypeCategory::Character, KIND>>;
+  using CharT = typename Character::value_type;
+
+public:
+  static constexpr bool IsValidCharacterCode(std::uint64_t code) {
+    if constexpr (KIND == 1) {
+      return IsValidASCII(code);
+    } else if constexpr (KIND == 2) {
+      return IsValidInternalEUC_JP(code);
+    } else if constexpr (KIND == 4) {
+      return IsValidUCS4(code);
+    } else {
+      static_assert(true, "bad character kind");
```

Comment:
Should this be static_assert(false...)?

Reply:
Yes, it should have been thanks. Removed this part as part of other comment below.